api/commands/run.ts

```diff
@@ -61,134 +83,223 @@ export class Runner {
+      this.runResult.actions.filter(
+        action => action.status !== dataform.ActionResult.ExecutionStatus.RUNNING
+      )
+    );
     const skippableActions = this.pendingActions.filter(allDependenciesHaveExecuted);
     this.pendingActions = this.pendingActions.filter(
       action => !allDependenciesHaveExecuted(action)
     );
+    skippableActions.forEach(skippableAction => {
+      this.runResult.actions.push({
+        name: skippableAction.name,
+        status: dataform.ActionResult.ExecutionStatus.SKIPPED,
+        tasks: skippableAction.tasks.map(() => ({
+          status: dataform.TaskResult.ExecutionStatus.SKIPPED
+        }))
+      });
+    });
+    const onActionsSkipped = async () => {
+      if (skippableActions.length > 0) {
+        await this.triggerChange();
+        await this.executeAllActionsReadyForExecution();
+      }
+    };
 
     await Promise.all([
+      onActionsSkipped(),
       Promise.all(
-        skippableActions.map(async skippableAction =>
-          this.onActionExecutionComplete(toSkippedAction(skippableAction))
-        )
-      ),
-      Promise.all(
-        executableActions.map(async executableAction =>
-          this.onActionExecutionComplete(await this.executeAction(executableAction))
-        )
+        executableActions.map(async executableAction => {
+          await this.executeAction(executableAction);
+          await this.executeAllActionsReadyForExecution();
+        })
       )
     ]);
   }
 
-  private async onActionExecutionComplete(executedAction: dataform.IExecutedAction) {
-    this.result.actions.push(executedAction);
-    await this.triggerChange();
-    await this.executeAllActionsReadyForExecution();
-  }
+  private async executeAction(action: dataform.IExecutionAction): Promise<void> {
+    if (action.tasks.length === 0) {
+      this.runResult.actions.push({
+        name: action.name,
+        status: dataform.ActionResult.ExecutionStatus.DISABLED,
+        tasks: []
+      });
+      await this.triggerChange();
+      return;
+    }
 
-  private async executeAction(action: dataform.IExecutionAction) {
-    const startTime = new Date().valueOf();
+    const timer = Timer.start();
+    const actionResult: dataform.IActionResult = {
+      name: action.name,
+      status: dataform.ActionResult.ExecutionStatus.RUNNING,
+      timing: timer.current(),
+      tasks: []
+    };
+    this.runResult.actions.push(actionResult);
+    await this.triggerChange();
 
-    const executedTasks: dataform.IExecutedTask[] = [];
-    let allSuccessful = true;
     for (const task of action.tasks) {
-      if (allSuccessful) {
-        const executedTask = await this.executeTask(task);
-        executedTasks.push(executedTask);
-        allSuccessful = allSuccessful && executedTask.ok;
+      if (this.cancelled) {
+        actionResult.tasks.push({
+          status: dataform.TaskResult.ExecutionStatus.CANCELLED
+        });
+        await this.triggerChange();
```

Comment:
this (and below) will trigger more often than it needs to I think? (ideally, we'd reduce the number of triggerChanges calls). As you are probably to skip many other actions and tasks.
This might be tricky to do so I'm not that bothered, but ideally triggerChange would fire only when:
- The moment a task finishes executing successfully
- After a task fails/cancels AND we mark everything else that can be as skipped

Reply:
Actually, because you trigger at the end of the loop and during execute task, if you remove this and below, I think that would be absolutely fine.